content/blog/2020-03-06-wasm/index.md

```diff
@@ -0,0 +1,229 @@
+classification overheads and overheads at parsing and verification.
+Existing technologies also has performance implications based on which browser 
+you target (different browsers used different techniques to optimize applications on
+browsers).
+
+## Is WebAssembly just C on browsers?
+WebAssembly does brings the performance of pre-compiled languages to web browsers. 
+But it does so side-by-side with traditional JavaScript running on the same engines.
+This allows browsers to have best of both worlds, drive the Ferrari on the race track--- 
+run performant WebAssembly when it needs performance--- or drive the Prius to the 
+grocery store--- quickly getting a simple web application up and running.
+
+With WebAssembly, you could use JavaScript for fast development, but also 
+use C where you need performance. You could use C (theoretically supports other languages too) where static typing is
+useful, but JavaScript where dynamic typing is a necessity for productivity.
+And you could combine these modules, leveraging interoperability to balance productivity and performance,
+as both these styles can now be executed in the same compiler flow.
+So WebAssembly is more than just C on browsers, it's a carefully coordinated, 
+well-designed, massive engineering effort to reconcile two worlds at odds.
+
+## How does it differ from existing tools?
+So WebAssembly is portable, but doesn't JVM (Java virtual machine) already 
+offer portability? Even though the overall goal has some similarities, Java 
+cannot be run on a browser without plugins and has doesn't have the language
+flexibilty WebAssembly provides. This is due to Java's portability coming 
+from the ability to run Java programs in different platforms, and therefore not
+designed to run multiple programs in a browser. Java also doesn't solve the issues
+JavaScript poses through interpreter based execution, dynamic typing and
+garbage collection. Moreover, validation and memory safety makes 
+WebAssembly more secure from vulnerabilities. Java was not designed with 
+formalization in mind, and therefore is complex to validate. Java applets
+have strict security restrictions as a work around for many implementation bugs.
+Structured control flow and single pass compilation and validation places WebAssembly as a 
+mature option compared to existing bytecode formats such as JVM and CIL 
+(Common Intermediate Language) which contain irreducible loops and unbalanced locking 
+features that are hard to be handled with JIT compilers and are relegated to 
+the interpreter. And these design decisions has made WebAssembly a better 
+candidate to implement performant and portable to run different languages 
+natively on a browser.
+
+WebAssembly also performs better compared to recent technologies with a similar
+goal. Firstly, it can outperform [asm.js][asmjs] from optimizations beyond what can be done with 
+JavaScript, WebAssembly has faster download with smaller code size, it does not require
```

Comment:
This wording doesn't quite parse: "from optimizations beyond what can be done with JS." And I am still confused about what you're saying here: that asm.js is "just JS" and therefore only gets JS-style optimizations? That's not true though—the point behind asm.js is that it gets to be optimized differently from JS (as you now note below).

Reply:
I want to say that even though `asm.js` has got very close to native performance, it is still limited by few aspects by having JS as the format, which WebAssembly has exploited to perform better than asm.js. I hope the description makes more sense now.

Reply:
I'm still not sure how to parse the phrase "it can outperform asm.js from optimizations beyond what can be done with JavaScript."